### PR TITLE
Fix batch-specific attention mod for NJT + Flex

### LINKED ISSUE #143788: Flex attention with nested tensors, bug in `create_nested_block_mask`
The following code is from `_nested_mod_func_adapter` which is a helper function used by `create_nested_block_mask`. Conceptually, it wraps a mod function that operates on individual batch items of a nested tensor and transforms the inputs so it works on a single packed item. However, the below code doesn't appear to update the batch argument (`b`) before calling the original mod function.

https://github.com/pytorch/pytorch/blob/6ccb8ed1868984d9d2ea4e48a085508d1027cd9b/torch/nn/attention/flex_attention.py#L985-L990

Since `create_nested_block_mask` effectively packs all the batch items from the nested tensor into a single item, it appears like the helper should do something like `b_nested = q_seq_idx[q_idx]` and pass that in place of `b` to `orig_mod_func`.

Otherwise, it appears that the wrapped mod func has no way of knowing which batch item it's operating on.

cc @cpuhrsch @jbschlosser @bhosmer @drisspg @soulitzer @davidberard98 @YuqingJ @chauhang @penguinwu @zou3519 @ydwu4 @bdhirsh @yf225 @Chillee @yanboliang @BoyuanFeng

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143866

Fixes #143788

### CODE REVIEW

File: torch/nn/attention/flex_attention.py
Comment (Skylion007) [inline torch/nn/attention/flex_attention.py:976]: Do we support inplace where yet for this tensor type? Seems like some free performance gains

**Comment by Skylion007** [comment]
This is also a good backport candidate for 2.6 since it's silent correctness.

**Comment by jbschlosser** [comment]
> This is also a good backport candidate for 2.6 since it's silent correctness.

Yeah true :( at the very least, this should make it into 2.6.1 if there is one and the release team doesn't want to cherry-pick it into the 2.6 release branch.

**Comment by Skylion007** [comment]
Just add the backport here once this PR lands: https://github.com/pytorch/pytorch/issues/142814 @jbschlosser

**Review by Skylion007 — APPROVED** [review]

File: torch/nn/attention/flex_attention.py
Comment (jbschlosser) [reply torch/nn/attention/flex_attention.py:976]: The tensor types here are:
* `q_seq_idx` - standard strided tensor
* `q_idx` - `BatchedTensor` due to vmap usage within flex

Not sure where inplace where would come in, but there is a [custom batching rule](https://github.com/pytorch/pytorch/blob/475656fd9c3871302e164bcd45415f18d68f637b/aten/src/ATen/functorch/BatchRulesIndexing.cpp#L17) for `index.Tensor`.

Comment (Skylion007) [reply torch/nn/attention/flex_attention.py:976]:
```suggestion
           output = orig_mod_func(score, b_nested, h, q_nested, kv_nested)
           output.where_(is_same_sequence, float("-inf"))
           return output
```

**Review by cpuhrsch — APPROVED** [review]
stamp

File: torch/nn/attention/flex_attention.py
Comment (jbschlosser) [reply torch/nn/attention/flex_attention.py:976]: Hm I tried this but it's actually not straightforward to use an in-place op here due to vmap:
```python
...
output = orig_mod_func(score, b_nested, h, q_nested, kv_nested)  # type: ignore[call-arg]
output.masked_fill_(q_seq_idx[q_idx] != kv_seq_idx[kv_idx], float("-inf"))
return output
```

```
RuntimeError: vmap: inplace arithmetic(self, *extra_args) is not possible because there exists a Tensor `other` in extra_args that has more elements than `self`. This happened due to `other` being vmapped over but `self` not being vmapped over in a vmap. Please try to use out-of-place operators instead of inplace arithmetic. If said operator is being called inside the PyTorch framework, please file a bug report instead.
```

gonna stick with the current logic for landing

**Pull request closed without merging**
